Add hooks for emitting lifetimes.

nate-chandler opened the pull request:
Add a new underscored attribute `@_eagerMove` (and also `@_lexical`) to indicate that a value or type should have an eager move lifetime (or a lexical lifetime if it would otherwise be eager move).

Update SILGen to only emit lexical borrow scopes for values which are not eager move.

Add these same attributes to `SILFunctionArgument`.  SILGen emits the lexical borrow scopes according to a variable's lifetime, but it doesn't emit lexical borrow scopes for arguments†.  In order for optimizations (CopyPropagation, SSADestroyHoisting, Inlining) to respect the lifetimes of arguments, the annotations those arguments received have to be available.

Updated CopyPropagation (indirectly, by way of `SILValue::isLexical`), SSADestroyHoisting, and SILInliner to respect these annotations.

† Currently lexical borrows are emitted for owned arguments, but that is superfluous and will be removed in an upcoming PR.

rdar://90411492

nate-chandler committed:
[NFC] Removed unused variable.

nate-chandler committed:
[Gardening] Tweaked comment grammar.

nate-chandler committed:
[SIL] Added SILType::isAggregate.

nate-chandler committed:
[SIL] Function args may have lifetime annotations.

By default, whether a function argument's lifetime is determined by its
type.  To allow that behavior to be overridden on an
argument-by-argument basis, the argument may have its own lifetime.

nate-chandler committed:
[Textual SIL] Explicit function arg lifetimes.

Print and parse @_eagerMove and @_lexical on function arguments.

nate-chandler committed:
[Serialization] Explicit function arg lifetimes.

nate-chandler committed:
[SILInliner] Respect explicit argument lifetimes.

nate-chandler committed:
[SILGen] Use decl for noImplicitCopy attr.

Rather than using the SILValue of the ManagedValue generated for that
argument, which is not always a SILFunctionArgument, just look directly
at the ParmDecl for the attribute.

nate-chandler committed:
[AST] Add simple LifetimeAnnotation enum.

The enum just provides a convenience around mutually exclusive
attributes which may be applied to decls and to SILFunctionArguments.
This way we can switch over the flavor of annotation that a decl enjoys
rather than having to remember the list of which might affect it, a list
which is likely to grow.

nate-chandler committed:
[SIL] Removed SILArgument::isNoImplicitCopy.

The attribute can't be applied to non-function arguments and may cause
confusion by being available on them.

nate-chandler committed:
[Sema] Added _eagerMove and _lexical attributes.

The new attributes affect how the lifetimes of values may be shortened.

nate-chandler committed:
[SIL] Added SILType::getLifetime.

The function describes the lifetime that variables of that type have by
default (unless otherwise annotated).  This is done by looking at leaf
nodes until one that is not eager move is found, at which point the
lifetime is known to be lexical.  Otherwise, the lifetime is eager move
(i.e. if every leaf is eager move).

nate-chandler committed:
[Gardening] Tweaked argument label.

nate-chandler committed:
[SILGen] Added TODO.

nate-chandler committed:
[SSADestroyHoisting] Lifetimes alter arg hoisting.

Arguments whose lifetimes are not lexical should be hoisted without
respect to deinit barriers.  On the other hand, inout arguments
explicitly annotated @_lexical should respect deinit barriers.

nate-chandler committed:
[SILGen] Only lexical types get lexical lifetimes.

Only emit `begin_borrow [lexical]` and only mark `alloc_stack`s
`[lexical]` when the variable in question's lifetime is lexical, not
eager move.

nate-chandler committed:
[SIL] Only lexical owned args are lexical.

Previously, every owned argument was considered lexical.  That included
owned arguments with eager move lifetimes.  Here, only owned arguments
whose lifetimes are lexical (i.e. not those whose lifetimes are eager
move) are considered lexical.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please benchmark

nate-chandler committed:
[TypeLowering] Record whether a type is lexical.

Used RecursiveProperties to keep track of whether a type is lexical.

This is done in the usual way by OR'ing together whether components are
lexicial.  The the values to be OR'd together come from primitives:
everything is non-lexical except reference types and resilient types
(because they could become reference types).  Those values can however
be affected by user annotations: marking a class @_lexical makes the
value for that primitive lexical; markinig an enum or struct @_eagerMove
overrides the value obtained by OR'ing together its components and makes
it be non-lexical.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
[MemAccessUtils] Non-trivial Cxx node is leaf.

Nodes which are non-trivial C++ decls must be destroyed, they cannot be
destroyed by way of destroying their subobjects.  This was exposed by
producing more non-lexical alloc_stacks which enjoy more aggressive
destroy_addr hoisting.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test windows platform

nate-chandler committed:
[SIL] Added simple Lifetime enum.

The new enum makes keeping track of what flavor of lifetime values ought
to have a little bit type-safer.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test source compatibility

atrick commented on docs/ReferenceGuides/UnderscoredAttributes.md:
```diff
@@ -131,6 +131,17 @@ library), instead of at an arbitrary point in time.
+that releases of the value may be hoisted without respect to deinit barriers.
+
+When applied to a type, indicates that all values which are instance of that
+type are themselves `@_eagerMove` as above, unless overridden with `@_lexical`.
```
We should be clear that `@_eagerMove` on types is only a hint. It means that the values may be eagerly moved, but will be honored only for concrete types. It's not broadly true for all "values which are instances of that type". It only applies to static types

nate-chandler replied:
Clarified that it only applies to values whose static type is eager move.

atrick commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3207,6 +3207,16 @@ ERROR(reasync_without_async_parameter,none,
+      "@_eagerMove requires lexical lifetimes to be enabled", ())
+
+ERROR(lexical_requires_lexical_lifetimes, none,
+      "@_lexical requires lexical lifetimes to be enabled", ())
```
Disabling lexical lifetimes in the compiler should just override attributes rather than breaking source. No need for a diagnostic.

nate-chandler replied:
Removed the diagnostic.

nate-chandler commented:
@swift-ci please test

atrick reviewed (approved):
Thanks!

Looks good on quick review. I'm assuming you caught all cases where the annotation needs to be checked.

It looks like you made sure that all non-trivial types have lexical semantics, unless there's an attribute. How confident are you and should we have verification to that effect?

Note that we really don't care if trivial types are marked lexical or not, so I wasn't actually expecting you to infer eager-move without an attribute, but it seems fine as long as we can verify that nothing falls through the cracks.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler committed:
[TypeLowering] Verify lexical property.

Check that if a type is non-trivial then either (1) it is lexical or (2)
its non-trivial leaves are @_eagerMove.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented on lib/SILGen/Scope.cpp:
```diff
@@ -57,6 +57,8 @@ static void lifetimeExtendAddressOnlyRValueSubValues(
     auto boxTy = SILBoxType::get(v->getType().getASTType());
     SILValue box = SGF.B.createAllocBox(loc, boxTy);
+    // TODO: Should these boxes that extend lifetimes for rvalue subobjects ever
+    //       be lexical?
```
Agreed--will do!

nate-chandler commented:
Added verification that a non-trivial type is either lexical or else all of its non-trivial aggregate "leaves" are eager move.

nate-chandler commented:
Source compat "failure" is a UPASS where the type checker no longer fails to complete quickly enough.

```
========================================
UPasses:
  UPASS: https://github.com/apple/swift/issues/60627, Doggie, 5.0, 965946, Swift Package
========================================
Action Summary:
     Passed: 336
     Failed: 0
    XFailed: 18
    UPassed: 1
      Total: 355
========================================
Repository Summary:
      Total: 115
========================================
Result: UPASS
========================================
```
